crates/common/tedge_utils/src/fs_notify.rs

```diff
@@ -0,0 +1,600 @@
+use std::{
+    collections::HashMap,
+    hash::Hash,
+    path::{Path, PathBuf},
+};
+
+use async_stream::try_stream;
+pub use futures::{pin_mut, Stream, StreamExt};
+use inotify::{EventMask, Inotify, WatchMask};
+use maplit::{btreeset, hashmap};
+use strum_macros::Display;
+use try_traits::default::TryDefault;
+
+const DIRECTORY_IS_WATCHED: &str = "C8Y_DIR_WATCH";
+
+#[derive(Debug, Display, PartialEq, Eq, Clone, Hash, PartialOrd, Ord)]
+pub enum FileEvent {
+    Modified,
+    Deleted,
+    Created,
+}
+
+impl TryFrom<FileEvent> for WatchMask {
+    type Error = NotifyStreamError;
+
+    fn try_from(value: FileEvent) -> Result<Self, Self::Error> {
+        match value {
+            FileEvent::Modified => Ok(WatchMask::MODIFY),
+            FileEvent::Deleted => Ok(WatchMask::DELETE),
+            FileEvent::Created => Ok(WatchMask::CREATE),
+        }
+    }
+}
+
+impl TryFrom<EventMask> for FileEvent {
+    type Error = NotifyStreamError;
+
+    fn try_from(value: EventMask) -> Result<Self, Self::Error> {
+        match value {
+            EventMask::MODIFY => Ok(FileEvent::Modified),
+            EventMask::DELETE => Ok(FileEvent::Deleted),
+            EventMask::CREATE => Ok(FileEvent::Created),
+            _ => Err(NotifyStreamError::UnsupportedEventMask { mask: value }),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum NotifyStreamError {
+    #[error(transparent)]
+    FromIOError(#[from] std::io::Error),
+
+    #[error("Error starting fs notification service.")]
+    NotifyInitError,
+
+    #[error("Error creating event stream")]
+    ErrorCreatingStream,
+
+    #[error("Error normalising watcher for: {path:?}")]
+    ErrorNormalisingWatcher { path: PathBuf },
+
+    #[error("Unsupported mask: {mask:?}")]
+    UnsupportedWatchMask { mask: WatchMask },
+
+    #[error("Unsupported mask: {mask:?}")]
+    UnsupportedEventMask { mask: EventMask },
+
+    #[error("Expected watch directory to be: {expected:?} but was: {actual:?}")]
+    WrongParentDirectory { expected: PathBuf, actual: PathBuf },
+
+    #[error("Watcher: {mask} is duplicated for file: {path:?}")]
+    DuplicateWatcher { mask: FileEvent, path: PathBuf },
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum Watch {
+    File(String),
+    Directory,
+}
+
+impl Watch {
+    fn eq_to_str(&self, value: &str) -> bool {
+        match &self {
+            Watch::File(file_name) => file_name.eq(value),
+            Watch::Directory => value.eq(DIRECTORY_IS_WATCHED),
```

Comment:
What exactly is this?
Why is the `value` compared to a constant in case `self` is `Watch::Directory`?

Maybe some documentation (here and at the constant definition) would help here!